File: app/src/main/java/dev/bee/kanjianki/MainActivitySettingsAnkiSourceNoteType.java
package dev.bee.kanjianki;

import android.widget.Toast;

import dev.bee.kanjianki.core.RecordsSyncModels;
import dev.bee.kanjianki.core.SettingsTextCopy;

final class MainActivitySettingsAnkiSourceNoteType {
    private final MainActivitySettings activity;

    MainActivitySettingsAnkiSourceNoteType(MainActivitySettings activity) {
        this.activity = activity;
    }

    android.view.View noteTypeSettingsPanel(RecordsSyncModels.Settings current) {
        RecordsSyncModels.Settings defaults = RecordsSyncModels.Settings.kikuDefaults();
        SettingsNoteTypeFieldState fieldState = new SettingsNoteTypeFieldState(
                current.modelName,
                current.expressionField,
                current.readingField,
                current.meaningField,
                current.sentenceField,
                current.frequencyField,
                current.frequencySortField
        );
        return MainActivitySettingsAnkiSourceNoteTypeCompose.noteTypeSettingsPanelView(
                activity,
                new SettingsNoteTypePanelModel(
                        SettingsTextCopy.noteTypeFieldsTitle(),
                        SettingsTextCopy.noteTypeUsingText(current.modelName),
                        SettingsTextCopy.noteTypeFieldsBody(),
                        fieldState,
                        SettingsTextCopy.requiredFieldsTitle(),
                        SettingsTextCopy.requiredFieldsBody(),
                        SettingsTextCopy.noteTypeStatusLabel(),
                        SettingsTextCopy.expressionFieldLabel(),
                        SettingsTextCopy.readingFieldLabel(),
                        SettingsTextCopy.meaningFieldLabel(),
                        SettingsTextCopy.sentenceFieldLabel(),
                        SettingsTextCopy.frequencyFieldLabel(),
                        SettingsTextCopy.frequencySortFieldLabel(),
                        SettingsTextCopy.chooseFromAnkiDroidLabel(),
                        SettingsTextCopy.useKikuLabel(),
                        SettingsTextCopy.saveNoteTypeLabel(),
                        () -> NoteTypeFieldMappings.choose(activity, activity.gateway, activity.io, activity.main, fieldState),
                        () -> fieldState.applyDefaults(defaults),
                        () -> saveNoteTypeFields(fieldState)
                )
        );
    }

    private void saveNoteTypeFields(SettingsNoteTypeFieldState fieldState) {
        String selectedNoteType = fieldState.getNoteType().trim();
        if (selectedNoteType.isEmpty()) {
            Toast.makeText(activity, SettingsTextCopy.noteTypeRequiredToast(), Toast.LENGTH_SHORT).show();
            return;
        }
        String selectedExpressionField = fieldState.getExpression().trim();
        if (selectedExpressionField.isEmpty()) {
            Toast.makeText(activity, SettingsTextCopy.expressionFieldRequiredToast(), Toast.LENGTH_SHORT).show();
            return;
        }
        SettingsWriteActions.saveNoteTypeFields(
                new SettingsWriteActions.NoteTypeFieldWriteRequest(
                        selectedNoteType,
                        selectedExpressionField,
                        fieldState.getReading().trim(),
                        fieldState.getMeaning().trim(),
                        fieldState.getSentence().trim(),
                        fieldState.getFrequency().trim(),
                        fieldState.getFrequencySort().trim()
                ),
                activity.store::putStringSetting
        );
        Toast.makeText(activity, SettingsTextCopy.noteTypeSavedToast(), Toast.LENGTH_LONG).show();
        activity.renderSettings();
    }
}

File: app/src/main/kotlin/dev/bee/kanjianki/MainActivitySettingsAnkiSourceNoteType.kt
package dev.bee.kanjianki

import android.widget.Toast
import dev.bee.kanjianki.core.RecordsSyncModels
import dev.bee.kanjianki.core.SettingsTextCopy

internal class MainActivitySettingsAnkiSourceNoteType(private val activity: MainActivitySettings) {
    fun noteTypeSettingsPanelModel(current: RecordsSyncModels.Settings): SettingsNoteTypePanelModel {
        val defaults = RecordsSyncModels.Settings.kikuDefaults()
        val fieldState = SettingsNoteTypeFieldState(
            current.modelName,
            current.expressionField,
            current.readingField,
            current.meaningField,
            current.sentenceField,
            current.frequencyField,
            current.frequencySortField
        )
        return SettingsNoteTypePanelModel(
            title = SettingsTextCopy.noteTypeFieldsTitle(),
            status = SettingsTextCopy.noteTypeUsingText(current.modelName),
            body = SettingsTextCopy.noteTypeFieldsBody(),
            fields = fieldState,
            requiredTitle = SettingsTextCopy.requiredFieldsTitle(),
            requiredBody = SettingsTextCopy.requiredFieldsBody(),
            noteTypeLabel = SettingsTextCopy.noteTypeStatusLabel(),
            expressionLabel = SettingsTextCopy.expressionFieldLabel(),
            readingLabel = SettingsTextCopy.readingFieldLabel(),
            meaningLabel = SettingsTextCopy.meaningFieldLabel(),
            sentenceLabel = SettingsTextCopy.sentenceFieldLabel(),
            frequencyLabel = SettingsTextCopy.frequencyFieldLabel(),
            frequencySortLabel = SettingsTextCopy.frequencySortFieldLabel(),
            chooseLabel = SettingsTextCopy.chooseFromAnkiDroidLabel(),
            kikuLabel = SettingsTextCopy.useKikuLabel(),
            saveLabel = SettingsTextCopy.saveNoteTypeLabel(),
            onChoose = SettingsNoteTypeAction {
                NoteTypeFieldMappings.choose(activity, activity.gateway, activity.io, activity.main, fieldState)
            },
            onUseKiku = SettingsNoteTypeAction { fieldState.applyDefaults(defaults) },
            onSave = SettingsNoteTypeAction { saveNoteTypeFields(fieldState) }
        )
    }

    private fun saveNoteTypeFields(fieldState: SettingsNoteTypeFieldState) {
        val selectedNoteType = fieldState.noteType.trim()
        if (selectedNoteType.isEmpty()) {
            Toast.makeText(activity, SettingsTextCopy.noteTypeRequiredToast(), Toast.LENGTH_SHORT).show()
            return
        }
        val selectedExpressionField = fieldState.expression.trim()
        if (selectedExpressionField.isEmpty()) {
            Toast.makeText(activity, SettingsTextCopy.expressionFieldRequiredToast(), Toast.LENGTH_SHORT).show()
            return
        }
        SettingsWriteActions.saveNoteTypeFields(
            SettingsWriteActions.NoteTypeFieldWriteRequest(
                selectedNoteType,
                selectedExpressionField,
                fieldState.reading.trim(),
                fieldState.meaning.trim(),
                fieldState.sentence.trim(),
                fieldState.frequency.trim(),
                fieldState.frequencySort.trim()
            ),
            activity.store::putStringSetting
        )
        Toast.makeText(activity, SettingsTextCopy.noteTypeSavedToast(), Toast.LENGTH_LONG).show()
        activity.renderSettings(true)
    }
}
